cypress-io/cypress
# feat: Component Testing banner

- Closes #26511 

### Additional details
_Note:_ The majority of files touched in this PR are just renaming the `loginConnectStore` to `userProjectStateStore`

Creating a new banner to help inform e2e users about component testing. This new banner will display under the following conditions:
1. No higher-priority banners to display (connection warnings, existing ACI banners, etc)
2. Cypress use > 4 days
3. At least 2 days since a previous banner of another type (if any) has been dismissed
4. At least 1 day since the Nav/CI popup has displayed (if it was previously shown)
5. Currently in e2e mode
6. CT is not configured (no `component` key in config)
7. At least one supported CT framework exists (anything auto-detected by the onboarding wizard)

### Steps to test
1. Create a new project, name it distinctively (eg `ct-banner`). Add dependencies that would satisfy at least one of our auto-detected frameworks - for example, `next` or `react` + `webpack`.
2. Open project in Cypress into e2e mode, scaffold example specs, and close Cypress.
3. Open Cypress app data location, find project's `state.json` file (on Mac this will live at `~/Library/Application\ Support/Cypress/cy/production/projects/{project}/state.json`, or use `Developer Tools > View App Data` menu item from launchpad). It should look like this:
```json
{
  "firstOpened": 1682966839869,
  "lastOpened": 1682966839869,
  "lastProjectId": null,
  "specFilter": ""
}
```
4. Modify the `firstOpened` timestamp to a value >= 4 days ago (easy way is to delete the first digit) and save
5. Reopen project into e2e mode - an ACI banner should display (which one depends on whether you're logged in - doesn't matter for this procedure).
	* Refresh the page, verify that banner still displays.
	* Close the banner using the "X".
	* Refresh and verify no banner displays.
	* Close Cypress
7. `state.json` file should now look like this - adjust the `dismissed` timestamp to a value >= 2 days ago
```json
{
  "firstOpened": 682966839869,
  "lastOpened": 1682966881904,
  "lastProjectId": null,
  "specFilter": "",
  "banners": {
    "aci_082022_connectProject": {
      "lastShown": 1682966943916,
      "dismissed": 682967079384
    }
  }
}
```
8. Reopen project into e2e mode, new CT Available banner should display
	* Banner should detect correct framework, use correct name & logo
	* `state.json` should now look like this:
```json
	{
  "firstOpened": 682966839869,
  "lastOpened": 1682967155424,
  "lastProjectId": null,
  "specFilter": "",
  "banners": {
    "aci_082022_connectProject": {
      "lastShown": 1682966943916,
      "dismissed": 682967079384
    },
    "ct_052023_available": {
      "lastShown": 1682967157281
    }
  }
}
```
9. Click "Quick Setup" - verify you are sent into the CT setup flow in launchpad. Go back to e2e mode
10. Click "Read the docs" - verify CT docs page is opened
11. Click "Not interested..." link
	* Verify survey is opened.
	* Verify that banner is now hidden in Cypress. Refresh page, verify banner is still not shown.
12. Close Cypress, edit `state.json` to remove `dismissed` key from `ct_052023_available` banner. Reopen project
13. Banner should once again show. Click "X" to dismiss, verify banner is hidden and does not reappear on page refresh.
14. Extra credit - by deleting the `dismissed` key and reopening you can validate the various conditions outlined above for when the banner should display/not display (CT not already configured, matching framework detected, etc)

### How has the user experience changed?

![Screenshot 2023-05-03 at 3 45 41 PM](https://user-images.githubusercontent.com/11482842/236045991-9110f663-11bf-4f94-9b16-07498cf587a7.png)

### PR Tasks


- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
